src/sagemaker_sklearn_extension/preprocessing/encoders.py

```diff
@@ -532,10 +557,34 @@ def fit(self, X, y=None):
         self
 
         """
-        # sklearn.preprocessing._BaseEncoder uses _categories due to deprecations in other classes
-        # can be removed once deprecations are removed
-        self._categories = self.categories
-        self._fit(X, handle_unknown="unknown")
+        if isinstance(self.max_categories, int) or self.threshold != 1:
+
+            assert self.max_categories >= 1
+            X_columns, n_samples, n_features = self._check_X(X)
+
+            if self.threshold == "auto":
+                threshold = max(10, n_samples / 1000)
+            elif self.threshold >= 1:
+                threshold = self.threshold
+            else:
+                threshold = ceil(self.threshold * n_samples)
+
+            for i in range(n_features):
+                dtype = X_columns[i].dtype
+                items, counts = np.unique(X_columns[i].astype(str), return_counts=True)
+                categories_to_encode = items[counts >= threshold]
+                if categories_to_encode.size == 0:
+                    warnings.warn(
+                        "feature at index {} does not have any categories appearing more than {} {}".format(
+                            i, threshold, "time" if threshold == 1 else "times"
+                        )
+                    )
+                if len(categories_to_encode) >= self.max_categories:
+                    most_freq_idxs = np.argsort(counts)[len(counts) - self.max_categories + 1 :]
```

Comment:
I think there's an off by one error here?

Notice how the fitted estimator doesn't have any categories set with `max_categories=1`
```
>>> X = [['Male', 1], ['Female', 3], ['Female', 1], ['Male', 2], ['Female', 3]]
>>> r = RobustOrdinalEncoder(max_categories=1)
>>> r.fit(X)
RobustOrdinalEncoder(categories=[array([], dtype=object),
                                 array([], dtype=object)],
                     max_categories=1)
```

But with `max_categories=2`, I get the right behavior
```
>>> X = [['Male', 1], ['Female', 3], ['Female', 1], ['Male', 2], ['Female', 3]]
>>> r = RobustOrdinalEncoder(max_categories=1)
>>> r.fit(X)
RobustOrdinalEncoder(categories=[array([], dtype=object),
                                 array([], dtype=object)],
                     max_categories=1)
```

Reply:
Fixing and adding a unit test.